### PR TITLE
Defer processing `ConfigValue`/`ConfigList` default value until first use

### LINKED ISSUE #3244: Defer processing ConfigDict's handling of default values until they are needed
## Summary

We lack an elegant way to pass configured solvers as config arguments, as is pointed out in #3220.

### Rationale

Many transformations rely on the user being able to pass a *configured* solver as an argument so that that configuration will persist wherever it is used in the transformation (or meta algorithm). Right now we often do that by passing a solver instance, but this is suboptimal since those solver instances will be created on import. (Alternatively, some parts of code take the solver name as the argument and pass an options dictionary separately, but this is rather clunky and difficult to validate.)

### Description

As suggested by @jsiirola, allow this syntax:
```python
CONFIG.declare(
    'solver',
    ConfigValue(
        domain=SolverFactory,
        default='gurobi',
        description="A solver to use to solve the continuous subproblems for "
        "calculating the M values",
    ),
)
```
and don't process the default value until the first time it is needed.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
**NOTE**: this PR builds on #3382.  You can see / review the new content in this PR through [this link](https://github.com/Pyomo/pyomo/compare/doc-reorg...jsiirola:config-default) 

It is useful to have `ConfigValue` objects that have complex (read: heavy-weight) domains.  For example, it would be convenient for a transformation to take a single configuration field for a (preconfigured) `solver`.  Unfortunately, we would like that domain to be a `SolverBase` instance - and providing a default solver would force that solver to be instantiated when the `ConfigValue` (`ConfigDict`) was *declared*.

This PR updated the internals to defer passing the `_default` value through `_cast` until the first time the `_data` is accessed / needed.

In addition, this cleans up some other issues in the Config classes:
- improve performance of serialization (avoid `dict`s)
- some classes were missing `__slots__` declarations
- update the `FlagType` so that instances of those classes can't be constructed (attempts to call the type will return the type and not an instance)

## Changes proposed in this PR:
- (see above)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/common/config.py [hunk 1]
```diff
@@ -1669,15 +1674,51 @@ def __call__(self, fcn):
         return fcn
 
 
+class UninitializedMixin(object):
```
Comment (mrmundt) [inline pyomo/common/config.py:1677]: Don't need `(object)` (we'll never stop having this argument, I know)

File: pyomo/common/config.py [hunk 2]
```diff
@@ -1669,15 +1674,51 @@ def __call__(self, fcn):
...
+    """Mixin class to support delayed data initialization.
+
+    This mixin can be used to create a derived Config class that hides
+    the (uninitialized) ``_data`` attribute behind a property.  Any
+    attempt to access the _data will trigger the initialization of the
+    Config object from its ``_default`` value.  Setting the ``_data``
+    attribute will also trigger resolution of the Config object, but
+    without processing the ``_default__.
```
Comment (mrmundt) [inline pyomo/common/config.py:1685]: Typo here probably?

File: pyomo/common/config.py [hunk 3]
```diff
@@ -2395,43 +2454,43 @@ def value(self, accessValue=True):
...
-        self._data.append(val)
-        self._data[-1]._parent = self
-        self._data[-1]._name = '[%s]' % (len(self._data) - 1,)
+        val._parent = self
+        val._name = f'[{len(_data)}]'
+        # We need to reset the _userSet to False because the List domain
+        # is a ConfigValue and __call__ will trigger set_value(), which
+        # will set the _userSet flag.  As we get here during _default
+        # processing, we want to clear that flag.  If this is actually
+        # getting triggeresd through set_value() / append(), then
```
Comment (mrmundt) [inline pyomo/common/config.py:2487]:
```suggestion
        # getting triggered through set_value() / append(), then
```

File: pyomo/common/config.py [hunk 4]
```diff
@@ -2508,8 +2572,8 @@ class ConfigDict(ConfigBase, Mapping):
 
     content_filters = {None, 'all', 'userdata'}
 
-    __slots__ = ('_declared', '_implicit_declaration', '_implicit_domain')
-    _all_slots = set(__slots__ + ConfigBase.__slots__)
+    __slots__ = ('_implicit_domain', '_declared', '_implicit_declaration')
```
Comment (mrmundt) [inline pyomo/common/config.py:2575]: Did the order matter?

File: pyomo/common/flags.py [hunk 5]
```diff
@@ -14,19 +14,31 @@
...
     fully-qualified class name (``__module__ + '.' + __qualname__``).
 
-    This is useful for defining "flag types" that are default arguments
-    in functions so that the Sphinx-generated documentation is "cleaner"
-
     """
 
+    def __new__(mcs, name, bases, dct):
+        # Ensure that attempts to construct instances of a Flag type
+        # returnthe type.
```
Comment (mrmundt) [inline pyomo/common/flags.py:35]:
```suggestion
        # return the type.
```

File: pyomo/common/tests/test_config.py [hunk 6]
```diff
@@ -1433,42 +1425,48 @@ def test_display_userdata_add_block_nonDefault(self):
 """,
         )
 
+    def test_display_nondata_type(self):
+        class NOOP(object):
```
Comment (mrmundt) [inline pyomo/common/tests/test_config.py:1429]: Do not need `(object)`

File: pyomo/common/tests/test_config.py [hunk 7]
```diff
@@ -2913,6 +2883,18 @@ def test_call_options(self):
         self.assertEqual(mod_copy._description, "new description")
         self.assertEqual(mod_copy._visibility, 0)
 
+    def test_template_nondata(self):
+        class NOOP(object):
```
Comment (mrmundt) [inline pyomo/common/tests/test_config.py:2887]: Do not need `(object)`

File: pyomo/common/tests/test_config.py [hunk 8]
```diff
@@ -3305,6 +3287,56 @@ def domain_name(self):
         cfg.declare('type', ConfigValue(domain=int))
         self.assertEqual(cfg.get('type').domain_name(), 'int')
 
+    def test_deferred_initialization(self):
+        class Accumulator(object):
```
Comment (mrmundt) [inline pyomo/common/tests/test_config.py:3291]: Do not need `(object)`

File: pyomo/common/tests/test_flags.py [hunk 9]
```diff
@@ -41,3 +41,9 @@ def test_NOTSET(self):
             del sys.modules['sphinx']
             in_testing_environment(None)
         self.assertIsNone(in_testing_environment.state)
+
+    def test_singleton(self):
+        # This tests that the type is a "singleton", and attempts to
+        # construct an instance will return the class
```
Comment (mrmundt) [inline pyomo/common/tests/test_flags.py:47]: Grammar weirdness here - maybe it's supposed to be this?
```suggestion
        # construct an instance that will return the class
```

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: pyomo/common/config.py (hunk 4 above)
Comment (jsiirola) [reply pyomo/common/config.py:2577]: Yes.  `__getstate__` processes all `__slots__` except for `_implicit_domain` in a general manner (`_implicit_domain` is handled separately).  This processing works by skipping the first element in `__slots__`.  See the Note on lines 2607-2608.

**Review by mrmundt — APPROVED** [review]

File: pyomo/common/config.py [hunk 10]
```diff
@@ -700,35 +704,37 @@ def from_enum_or_string(cls, arg):
...
+    ))
+    config.declare("iteration limit", ConfigValue(
+        default=30,
+        domain=int,
+        description="Iteration limit",
+        doc="Number of maximum iterations in the decomposition methods"
+    ))
+
+Users can then to provide values for those entries, and retrieve the
+current values:
```
Comment (emma58) [inline pyomo/common/config.py:734]:
```suggestion
Users can then provide values for those entries, and retrieve the
current values:
```

File: pyomo/common/tests/test_flags.py [hunk 11]
```diff
@@ -41,3 +41,9 @@ def test_NOTSET(self):
             del sys.modules['sphinx']
             in_testing_environment(None)
         self.assertIsNone(in_testing_environment.state)
+
+    def test_singleton(self):
+        # This tests that the type is a "singleton", and that any
+        # attempts to construct an instance will return the class
+        self.assertIs(NOTSET(), NOTSET)
+        self.assertIs(NOTSET(), NOTSET())
```
Comment (emma58) [inline pyomo/common/tests/test_flags.py:49]: This is sick. :P

**Review by emma58 — APPROVED** [review]
Wheeeeeeeee!

**Pull request merged**